### PR TITLE
Added JSON verification functionality for the Configure Name Dialog box

### PR DESCRIPTION
**Pull request by junekomeiji**
I added JSON verification functionality in names/applications.cpp alongside the associated unit tests and added them to the configure name dialog box. The dialog box should tell you if you inputted an invalid JSON value and warn you against configuring your domain with said invalid value, but should allow you to continue anyway if you were to insist.

### CODE REVIEW

**Comment by JeremyRand** [comment]
Hi Rose! Good to see your first PR here.

Can you squash the Git history down to one commit? (Happy to review the code without blocking on that, but it'll need to be squashed before a merge can happen.)

@domob1812 If you want me to do a first pass at review before you look at it, that's OK with me (seeing as this PR is to a spec that I provided).

File: src/names/applications.cpp
Comment (JeremyRand) [inline src/names/applications.cpp:187]: This log message was useful during testing but I think we can remove it now that the code is tested to work. Otherwise it'll spam logs a lot (I guess one message per keystroke in this textbox?).

File: src/names/applications.cpp
Comment (JeremyRand) [inline src/names/applications.cpp:11]: If the log message is removed below, I guess this #include can also be removed?

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [inline src/qt/configurenamedialog.cpp:56]: Is there a reason not to use `returnData`? (Maybe there is, it just jumps out at me as a plausible alternative.)

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [inline src/qt/configurenamedialog.cpp:70]: Maybe tweak this message a bit to more explicitly indicate that continuing is likely to make the name unresolvable?

File: src/qt/configurenamedialog.cpp
Comment (junekomeiji) [reply src/qt/configurenamedialog.cpp:56]: I did that and it doesn't work anymore, no idea why

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [inline src/qt/configurenamedialog.cpp:69]: Per [this Microsoft UI guidance](https://learn.microsoft.com/en-us/windows/win32/uxguide/vis-std-icons), I think a warning icon should probably be used rather than a question icon. It's possible I'm misreading the guidance, let me know if you disagree.

File: src/qt/configurenamedialog.cpp
Comment (junekomeiji) [reply src/qt/configurenamedialog.cpp:56]: By it I mean the dialogue box, again no idea why

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [inline src/qt/configurenamedialog.cpp:118]: I suspect that showing text in the label when the JSON is valid will be distracting. Would it make sense to set the label to the empty string in this case instead?

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [inline src/qt/configurenamedialog.cpp:120]: The "minimal" check is being saved for the next PR, right?

Comment (junekomeiji) [reply src/qt/configurenamedialog.cpp:120]: Yeah, that is correct

File: src/qt/configurenamedialog.cpp
Comment (junekomeiji) [reply src/qt/configurenamedialog.cpp:118]: Makes sense

File: src/qt/configurenamedialog.cpp
Comment (junekomeiji) [reply src/qt/configurenamedialog.cpp:69]: Makes sense, it is a warning after all

File: src/test/name_applications_tests.cpp
Comment (JeremyRand) [inline src/test/name_applications_tests.cpp:127]: Typo, `a invalid` should be `an invalid`

File: src/qt/configurenamedialog.cpp
Comment (junekomeiji) [reply src/qt/configurenamedialog.cpp:70]: Sure

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [inline src/qt/configurenamedialog.cpp:117]: There's an edge case here. If the user enters the empty string, I'm guessing that will be marked as invalid JSON (am I correct?). It's not really ideal for the user in this scenario to be encouraged to replace it with an empty JSON object `{}` since that will waste 2 bytes of blockchain space for no reason. Can we explicitly whitelist the empty string so that it doesn't produce a warning?

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [reply src/qt/configurenamedialog.cpp:118]: Actually hmm, maybe this actually is beneficial, in the sense that if we later add support for CBOR as a more compact alternative to JSON, it could actually be informative. I'm going to tentatively retract this request.

Comment (JeremyRand) [reply src/qt/configurenamedialog.cpp:118]: Actually I'll leave it up to you. :) I don't feel super strongly here.

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [reply src/qt/configurenamedialog.cpp:56]: You did run `.toStdString()` on `returnData` when trying it, right? If that doesn't work, then I won't worry about why, just was curious.

**Comment by JeremyRand** [comment]
Reviewed 124075a9fac65a650bf3e7ea3827e943c099659b (didn't try to compile or test it yet): overall this looks very solid. Requesting some small tweaks (or clarifications on rationale) in the above inline review, but I don't see any major concerns based on looking at the code. I'll look at compiling and running it in the next hour or two, but I don't anticipate any negative surprises when I do so.

**Comment by JeremyRand** [comment]
Built 124075a9fac65a650bf3e7ea3827e943c099659b from source. Functionality looks good except for one oddity: clicking OK in the warning dialog (when it's invalid JSON) results in the expected behavior (transaction is issued), but if the warning dialog isn't shown (because it's valid JSON), clicking the OK button in the Configure Name dialog doesn't seem to have any effect at all.

I'm not 100% confident that this is a real bug, it might be something related to my wallet only having unconfirmed inputs (had to set up a new wallet for $REASONS), but since it works when the warning is shown, I suspect the bug may be real. Rose, can you check if this behavior is reproducible on your end?

File: src/qt/configurenamedialog.cpp
Comment (domob1812) [inline src/qt/configurenamedialog.cpp:67]: That `|` should be `||` (logical OR instead of bitwise OR).

I personally would also prefer `!data.empty ()` over `data != ""`, but that is a matter of style and both are fine with me.

Generally speaking though, what exactly is the intent of the condition?  I would find it logical if it was `!IsValidJSON(data) && !data.empty()` instead of OR, i.e. the "accepted" data is valid JSON or the empty string (since that might be a common use of invalid JSON).

File: src/qt/configurenamedialog.cpp
Comment (domob1812) [inline src/qt/configurenamedialog.cpp:115]: Here again, please replace `|` by `||`.

File: src/test/name_applications_tests.cpp
Comment (domob1812) [inline src/test/name_applications_tests.cpp:129]: The test here doesn't matter too much, since the main logic is from Univalue anyway (and so you are just testing whether or not you use Univalue correctly to parse JSON).

I do find it distracting that such a complex value is used, though.  For your purposes, you don't need a real value from Namecoin at all, and could use something very simple like `null`, `false`, `"foo"` or `{"bar": [1, 2, 3]}` as valid values, and `{"foo":` as invalid (or many other simple strings obviously).

**Review by domob1812 — COMMENTED** [review]
Thanks for the PR!  I'm not a Qt expert so I leave that bit to testing and Jeremy, but I have added some general comments on the code.

**Comment by junekomeiji** [comment]
@domob1812 's comments resolved the bug Jeremy was talking about, so thank you for that! I've also fixed the stuff you've talked about. Anything else preventing merging for now?

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [reply src/qt/configurenamedialog.cpp]: Since this logic shows up twice in this file, would it be OK to move the empty-string special-case logic into the `IsValidJSON` function? I realize this is maybe semantically misleading since an empty string is not actually valid JSON, but since that function lives in `applications.cpp` and I can't think of literally any contexts where application-layer values that need a JSON object would need to know the difference between an empty JSON object an an empty string, I think this would reduce the risk of accidents if we forget to copy over the special-case logic into future code that calls that function.

File: src/test/name_applications_tests.cpp
Comment (JeremyRand) [reply src/test/name_applications_tests.cpp:127]: I prefer a JSON object rather than a non-object JSON value such as a string or boolean, but other than that, yes this is a reasonable request.

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [inline src/qt/configurenamedialog.cpp:12]: Do we still need `logging.h` here?

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [reply src/qt/configurenamedialog.cpp:69]: I think this is not resolved yet? Or am I misreading GitHub's UI?

File: src/qt/configurenamedialog.cpp
Comment (JeremyRand) [inline src/qt/configurenamedialog.cpp:120]: The lack of a space before `else` is visually distracting to me, but I'll let @domob1812 decide what the appropriate style is here.

**Comment by JeremyRand** [comment]
Added a little more feedback inline -- this is progressing nicely. Let me know when the last bit of feedback is addressed, then I'll do a final build-and-test.

**Comment by JeremyRand** [comment]
I think all that remains in terms of code review is changing `question` to `warning`?

**Comment by JeremyRand** [comment]
5be985b49988df9ff8a8ebadb03c5ee54149d650 looks OK to me code-wise at this point; will do some testing in the next few hours to confirm.

File: src/qt/forms/configurenamedialog.ui
Comment (JeremyRand) [inline src/qt/forms/configurenamedialog.ui:195]: I think Qt Designer ate this XML comment when you changed the layout and moved its parent element to lines 64-84. Can you re-add this comment via manual text editing? (I don't know of a way to make Qt Designer not eat them, yes I know it's annoying.)

**Comment by JeremyRand** [comment]
Successfully tested 5be985b49988df9ff8a8ebadb03c5ee54149d650, functionality looks good to me. At this point all that we need for me to recommend that Daniel merge is fixing the XML comment issue I just noticed, and one last Git squash down to one commit.

File: src/test/name_applications_tests.cpp
Comment (JeremyRand) [inline src/test/name_applications_tests.cpp:122]: Now that the empty string is special-cased in the implementation, can we have a test case here for that?

**Comment by JeremyRand** [comment]
> Successfully tested [5be985b](https://github.com/namecoin/namecoin-core/commit/5be985b49988df9ff8a8ebadb03c5ee54149d650), functionality looks good to me. At this point all that we need for me to recommend that Daniel merge is fixing the XML comment issue I just noticed, and one last Git squash down to one commit.

Well, that plus the extra test case I just requested above.

File: src/test/name_applications_tests.cpp
Comment (JeremyRand) [reply src/test/name_applications_tests.cpp:122]: Oh wait, I just realized that special case didn't get moved like I suggested. Can that be moved? I think it would make the code cleaner.

**Comment by JeremyRand** [comment]
ACK 83cd7b2df91223bb7c34cff662b4b9748aa78434. Code review looks fine, unit tests pass, and functionality works as expected. Nice work Rose!

@domob1812 Feel free to merge this if you're satisfied that your review has been addressed.

File: src/qt/configurenamedialog.cpp
Comment (domob1812) [reply src/qt/configurenamedialog.cpp]: Yes that makes sense to me.  Maybe call the function then `IsValidJSONOrEmpty` just to avoid any confusion.

File: src/test/name_applications_tests.cpp
Comment (domob1812) [reply src/test/name_applications_tests.cpp:127]: Well, that depends on what the logic should be.  If any JSON value should be acceptable, then I think the test should probably explicitly run both object and non-object values through the function.  If we only want to accept JSON objects, then obviously the logic itself should be updated (we have exactly the same validation, although on the consensus level, in Xaya btw), and the test verify that, too.

File: src/qt/configurenamedialog.cpp
Comment (domob1812) [reply src/qt/configurenamedialog.cpp]: Yes I think there should be spaces around `else`, and also a space after `if` and between `){` at the end of the first line.  But the style in the codebase is a bit mixed up anyway, so I won't block changes for style in Qt code.

File: src/names/applications.cpp
Comment (domob1812) [inline src/names/applications.cpp:182]: Please consider renaming this to `IsValidJSONOrEmptyString` as per my other comment.  Doesn't matter much, but will avoid any potential confusion in the future.

File: src/names/applications.cpp
Comment (domob1812) [inline src/names/applications.cpp:185]: I would change this to `test.empty() || v.read(text)`, since checking for the empty string is a simpler operation than JSON-parsing (even though JSON parsing of the empty string is probably also easy), and thus the shortcut makes more sense.

**Review by domob1812 — APPROVED** [review]
Here are two more minor comments.  I'm fine either way, though, and happy to merge also if you prefer not to address them.

**Comment by junekomeiji** [comment]
I've addressed everything, it's time to merge I think

**Comment by JeremyRand** [comment]
ACK 9bd6e0ad145ba74e2ddb52d1be0bebdd4cdae0b4. Code looks fine, unit tests pass, manual Qt testing shows no issues.

@domob1812 Feel free to merge.

File: src/test/name_applications_tests.cpp
Comment (JeremyRand) [reply src/test/name_applications_tests.cpp:127]: My rough vision is that a subsequent PR would add some validation on specific JSON fields, and that when we do that, restricting to a JSON object would make sense (at which point we'd add more tests accordingly). I don't see any reason to tie that to this PR (incrementalism is good), but I also don't see any reason to explicitly test behavior that's only temporary and which will be changed (and which no users are relying on).

**Comment by domob1812** [comment]
ACK 9bd6e0ad145ba74e2ddb52d1be0bebdd4cdae0b4.

**Pull request merged**
